rust-src/bulletproofs/src/set_non_membership_proof.rs

```diff
@@ -0,0 +1,674 @@
+/// - `V` - commitment to `v`
+/// - `proof` - the set-non-membership proof to verify
+/// - `gens` - generators containing vectors `G` and `H` both of length at least
+///   `k` where k is the smallest power of two >= `|the_set|` (bold **g**,**h**
+///   in bluepaper)
+/// - `v_keys` - commitment keys `B` and `B_tilde` (`g,h` in bluepaper)
+#[allow(non_snake_case)]
+pub fn verify<C: Curve>(
+    transcript: &mut RandomOracle,
+    the_set: &[C::Scalar],
+    V: &Commitment<C>,
+    proof: &SetNonMembershipProof<C>,
+    gens: &Generators<C>,
+    v_keys: &CommitmentKey<C>,
+) -> Result<(), VerificationError> {
+    // Part 1: Setup
+    // Pad set if not power of two
+    let mut set_vec = the_set.to_vec();
+    pad_vector_to_power_of_two(&mut set_vec);
+    let n = set_vec.len();
+    if gens.G_H.len() < n {
+        return Err(VerificationError::NotEnoughGenerators);
+    }
+    // Select generators for vector commitments
+    let (G, H): (Vec<_>, Vec<_>) = gens.G_H.iter().take(n).cloned().unzip();
+
+    // Domain separation
+    transcript.add_bytes(b"SetNonMembershipProof");
+    // append commitment V to transcript
+    transcript.append_message(b"V", &V.0);
+    transcript.append_message(b"theSet", &set_vec);
+
+    // define the commitments A,S
+    let A = proof.A;
+    let S = proof.S;
+    // append commitments A and S to transcript
+    transcript.append_message(b"A", &A);
+    transcript.append_message(b"S", &S);
+
+    // get challenges y,z from transcript
+    let y: C::Scalar = transcript.challenge_scalar::<C, _>(b"y");
+    let z: C::Scalar = transcript.challenge_scalar::<C, _>(b"z");
+
+    // define the commitments T1, T2
+    let T_1 = proof.T_1;
+    let T_2 = proof.T_2;
+    // append T1, T2 commitments to transcript
+    transcript.append_message(b"T1", &T_1);
+    transcript.append_message(b"T2", &T_2);
+
+    // get challenge x (evaluation point) from transcript
+    let x: C::Scalar = transcript.challenge_scalar::<C, _>(b"x");
+
+    // define polynomial evaluation value
+    let tx = proof.tx;
+    // define blinding factors for tx and IP proof
+    let tx_tilde = proof.tx_tilde;
+    let e_tilde = proof.e_tilde;
+    // append tx, tx_tilde, e_tilde to transcript
+    transcript.append_message(b"tx", &tx);
+    transcript.append_message(b"tx_tilde", &tx_tilde);
+    transcript.append_message(b"e_tilde", &e_tilde);
+
+    // get challenge w from transcript
+    let w: C::Scalar = transcript.challenge_scalar::<C, _>(b"w");
+
+    // compute delta(y,z) = <1, y^n>  - z<s, y^n>
+    let yn = z_vec(y, 0, n);
+    let one_vec = vec![C::Scalar::one(); n];
+    let ip_one_yn = inner_product(&one_vec, &yn);
+    let mut zip_s_yn = inner_product(&set_vec, &yn);
+    zip_s_yn.mul_assign(&z);
+    let mut delta_yz = ip_one_yn;
+    delta_yz.sub_assign(&zip_s_yn);
+
+    // Part 2: Verify consistency of t_0
+    // i.e., check 0 = V^(z <1, yn>) * g^(delta_yz - tx) * T_1^x * T_2^x^2 *
```

Comment:
with multiplicative notation, it should be '1' on the LHS; but that would also make the reader confused!

Reply:
Yeah. This also applies to the other proofs. We could say instead 1_G as in the bluepaper or write verbosely "check that ... is neutral element".